### PR TITLE
Adds USB port to Fire Alarms

### PR DESCRIPTION
**Pull request by Oricana-16**
## About The Pull Request

Gives fire alarms a usb port that lets you toggle it on and off and check its state

## Why It's Good For The Game

Buffs circuits to interact with the world more, while also not making them better at killing people

## Changelog

:cl:
add: Added a usb port to fire alarms
/:cl:

### CODE REVIEW

**Comment by Oricana-16** [comment]
> firelocks, which can crush and hurt people
fire locks are so slow that, while it technically would hurt you need to not be paying attention for it to hurt. unless you do make a trap like you said which imo just makes circuits better bc you have to be smart about it.

**Comment by Watermelon914** [comment]
> This theoretically makes them better at killing, because fire alarms control firelocks, which can crush and hurt people. could be fun for a trap, though.

this is not an issue, since it's not directly a weapon. People are allowed to make unique and fun weapons with circuits, it's encouraged as an antagonist

File: code/game/machinery/firealarm.dm [hunk 1]
```diff
@@ -176,6 +180,7 @@
 	if(user)
 		log_game("[user] triggered a fire alarm at [COORD(src)]")
 	soundloop.start() //Manually pulled fire alarms will make the sound, rather than the doors.
+	SEND_SIGNAL(src, COMSIG_FIREALARM_USED)
```
Comment (Watermelon914) [inline code/game/machinery/firealarm.dm:183]: use two separate signals for this, don't call the same signal in multiple different places.
e.g.
COMSIG_FIREALARM_ON_TRIGGER
COMSIG_FIREALARM_ON_RESET

File: code/game/machinery/firealarm.dm [hunk 2]
```diff
@@ -419,3 +425,37 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/firealarm, 26)
...
+	var/obj/machinery/firealarm/attached_alarm
+
+/obj/item/circuit_component/firealarm/populate_ports()
+	is_on = add_output_port("Active", PORT_TYPE_NUMBER)
+
+/obj/item/circuit_component/firealarm/register_usb_parent(atom/movable/parent)
+	. = ..()
+	if(istype(parent, /obj/machinery/firealarm))
+		attached_alarm = parent
+		RegisterSignal(parent, COMSIG_FIREALARM_USED, .proc/on_firealarm_used)
```
Comment (Watermelon914) [inline code/game/machinery/firealarm.dm:446]: if you have two separate signals here, you don't need an if statement.

File: code/game/machinery/firealarm.dm [hunk 3]
```diff
@@ -419,3 +425,37 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/firealarm, 26)
 	if (!party_overlay)
 		party_overlay = iconstate2appearance('icons/turf/areas.dmi', "party")
 	area.add_overlay(party_overlay)
+
+/obj/item/circuit_component/firealarm
+	display_name = "Fire Alarm"
+	desc = "Allows you to interface with the Fire Alarm."
+	circuit_flags = CIRCUIT_FLAG_INPUT_SIGNAL
+
+	/////When the firealarm is used
```
Comment (Watermelon914) [inline code/game/machinery/firealarm.dm:434]:
```suggestion
	/// Returns a boolean value of 0 or 1 if the fire alarm is on or not.
```

**Comment by Mothblocks** [comment]
@datorangebottle 

Non-goal #3:

> Provide weaponry. While it is completely fine, and even encouraged, for players to use circuits for antagonistic purposes, **these should come from designs that make clever use of their environment (such as a circuit that locks people in a room and gases them)** rather than unimaginative weapon components.

File: code/game/machinery/firealarm.dm (hunk 2 above)
Comment (Mothblocks) [reply code/game/machinery/firealarm.dm:446]: Agree

**Comment by Oricana-16** [comment]
actually I think it would've been better to request moth cause I have no idea if watermelon is able to merge or do anything like that. my b.

**Comment by Mothblocks** [comment]
He is

**Comment by TheBonded** [comment]
I wonder if it would be possible to add usb connections to wall mounted buttons too? then you could remotely control blast doors

Unless the assembly shell already does this

**Comment by Squalhardt** [comment]
you'd need to check if it can already be used with assembly shells, but otherwise yeah its very possible.

**Pull request merged**